File: cucumber-expressions/go/cucumber_expression_tokenizer.go
package cucumberexpressions

func tokenize(expression string) ([]token, error) {

	tokens := make([]token, 0)

	runes := []rune(expression)

	var buffer []rune
	previousTokenType := startOfLine
	treatAsText := false
	escaped := 0
	bufferStartIndex := 0

	convertBufferToToken := func(tokenType tokenType) token {
		escapeTokens := 0
		if tokenType == text {
			escapeTokens = escaped
			escaped = 0
		}
		consumedIndex := bufferStartIndex + len(buffer) + escapeTokens
		t := token{string(buffer), tokenType, bufferStartIndex, consumedIndex}
		buffer = []rune{}
		bufferStartIndex = consumedIndex
		return t
	}

	tokenTypeOf := func(r rune, treatAsText bool) (tokenType, error) {
		if !treatAsText {
			return typeOf(r)
		}
		if canEscape(r) {
			return text, nil
		}
		return startOfLine, createCantEscaped(expression, bufferStartIndex+len(buffer)+escaped)
	}

	shouldCreateNewToken := func(currentTokenType tokenType, previousTokenType tokenType) bool {
		if currentTokenType != previousTokenType {
			return true
		}
		return currentTokenType != whiteSpace && currentTokenType != text
	}

	if len(runes) == 0 {
		tokens = append(tokens, token{"", startOfLine, 0, 0})
	}

	for _, r := range runes {
		if !treatAsText && isEscapeCharacter(r) {
			escaped++
			treatAsText = true
			continue
		}

		currentTokenType, err := tokenTypeOf(r, treatAsText)
		if err != nil {
			return nil, err
		}
		treatAsText = false

		if shouldCreateNewToken(currentTokenType, previousTokenType) {
			token := convertBufferToToken(previousTokenType)
			previousTokenType = currentTokenType
			buffer = append(buffer, r)
			tokens = append(tokens, token)
		} else {
			previousTokenType = currentTokenType
			buffer = append(buffer, r)
		}
	}

	if len(buffer) > 0 {
		token := convertBufferToToken(previousTokenType)
		tokens = append(tokens, token)
	}

	if treatAsText {
		return nil, createTheEndOfLineCanNotBeEscaped(expression)
	}
	token := token{"", endOfLine, len(runes), len(runes)}
	tokens = append(tokens, token)
	return tokens, nil
}

File: cucumber-expressions/javascript/src/CucumberExpressionTokenizer.ts
import { Token, TokenType } from './Ast'
import { createCantEscaped, createTheEndOfLIneCanNotBeEscaped } from './Errors'

export default class CucumberExpressionTokenizer {
  tokenize(expression: string): readonly Token[] {
    const codePoints = Array.from(expression)
    const tokens: Array<Token> = []
    let buffer: Array<string> = []
    let previousTokenType = TokenType.startOfLine
    let treatAsText = false
    let escaped = 0
    let bufferStartIndex = 0

    function convertBufferToToken(tokenType: TokenType): Token {
      let escapeTokens = 0
      if (tokenType == TokenType.text) {
        escapeTokens = escaped
        escaped = 0
      }

      const consumedIndex = bufferStartIndex + buffer.length + escapeTokens
      const t = new Token(tokenType, buffer.join(''), bufferStartIndex, consumedIndex)
      buffer = []
      bufferStartIndex = consumedIndex
      return t
    }

    function tokenTypeOf(codePoint: string, treatAsText: boolean): TokenType {
      if (!treatAsText) {
        return Token.typeOf(codePoint)
      }
      if (Token.canEscape(codePoint)) {
        return TokenType.text
      }
      throw createCantEscaped(expression, bufferStartIndex + buffer.length + escaped)
    }

    function shouldCreateNewToken(previousTokenType: TokenType, currentTokenType: TokenType) {
      if (currentTokenType != previousTokenType) {
        return true
      }
      return currentTokenType != TokenType.whiteSpace && currentTokenType != TokenType.text
    }

    if (codePoints.length == 0) {
      tokens.push(new Token(TokenType.startOfLine, '', 0, 0))
    }

    codePoints.forEach((codePoint) => {
      if (!treatAsText && Token.isEscapeCharacter(codePoint)) {
        escaped++
        treatAsText = true
        return
      }
      const currentTokenType = tokenTypeOf(codePoint, treatAsText)
      treatAsText = false

      if (shouldCreateNewToken(previousTokenType, currentTokenType)) {
        const token = convertBufferToToken(previousTokenType)
        previousTokenType = currentTokenType
        buffer.push(codePoint)
        tokens.push(token)
      } else {
        previousTokenType = currentTokenType
        buffer.push(codePoint)
      }
    })

    if (buffer.length > 0) {
      const token = convertBufferToToken(previousTokenType)
      tokens.push(token)
    }

    if (treatAsText) {
      throw createTheEndOfLIneCanNotBeEscaped(expression)
    }

    tokens.push(new Token(TokenType.endOfLine, '', codePoints.length, codePoints.length))
    return tokens
  }
}
